onnx/defs/schema.cc

```diff
@@ -667,55 +700,186 @@ void OpSchema::ParseAndSetTypes(
+
+  // opset import may have been set
+  // we may need to update its version with the specified opset_version
+  UpdateFunctionProtoOpsetImportVersion(*function_proto, opset_version);
+
+  opset_version_to_function_body_.insert(std::make_pair(opset_version, function_proto));
   return *this;
 }
 
-OpSchema& OpSchema::FunctionBody(const std::vector<NodeProto>& func_nodes) {
+OpSchema& OpSchema::FunctionBody(const std::vector<NodeProto>& func_nodes, int opset_version) {
+  if (opset_version == OpSchema::kUninitializedSinceVersion && since_version_ != OpSchema::kUninitializedSinceVersion) {
+    opset_version = since_version_;
+  }
+  std::shared_ptr<FunctionProto> function_proto(new FunctionProto());
   for (const auto& node : func_nodes) {
-    auto new_node = function_body_.add_node();
+    auto new_node = function_proto->add_node();
     new_node->CopyFrom(node);
   }
+
+  // opset import may have been set
+  // we may need to update its version with the specified opset_version
+  UpdateFunctionProtoOpsetImportVersion(*function_proto, opset_version);
+  opset_version_to_function_body_.insert(std::make_pair(opset_version, function_proto));
   return *this;
 }
 
 OpSchema& OpSchema::FunctionBody(
     const std::vector<NodeProto>& func_nodes,
-    const std::vector<OperatorSetIdProto>& relied_opsets) {
-  for (auto& relied_opset : relied_opsets) {
-    *(function_body_.mutable_opset_import()->Add()) = relied_opset;
+    const std::vector<OperatorSetIdProto>& relied_opsets,
+    int opset_version) {
+  if (opset_version == OpSchema::kUninitializedSinceVersion && since_version_ != OpSchema::kUninitializedSinceVersion) {
+    opset_version = since_version_;
   }
 
-  return FunctionBody(func_nodes);
-}
+  std::shared_ptr<FunctionProto> function_proto(new FunctionProto());
+  for (auto& relied_opset : relied_opsets) {
+    *(function_proto->mutable_opset_import()->Add()) = relied_opset;
+  }
 
-OpSchema& OpSchema::FunctionAddOpset(const char* domain, int version) {
-  OperatorSetIdProto* onnx_opset = function_body_.mutable_opset_import()->Add();
-  onnx_opset->set_domain(domain);
-  onnx_opset->set_version(version);
+  for (const auto& node : func_nodes) {
+    auto new_node = function_proto->add_node();
+    new_node->CopyFrom(node);
+  }
+  // opset import may have been set
+  // we may need to update its version with the specified opset_version
+  UpdateFunctionProtoOpsetImportVersion(*function_proto, opset_version);
+  opset_version_to_function_body_.insert(std::make_pair(opset_version, function_proto));
   return *this;
 }
 
 const FunctionProto* OpSchema::GetFunction() const {
-  return function_body_.node_size() > 0 ? &function_body_ : nullptr;
+  return GetFunctionWithOpsetVersion(since_version_);
+}
+
+const FunctionProto* OpSchema::GetFunctionWithOpsetVersion(int requested_opset_version) const {
+  return const_cast<OpSchema*>(this)->GetFunctionWithOpsetInternal(requested_opset_version);
```

Comment:
Why do we need this `const_cast` ? May be outdated. I don't see any UpdateFunctionProtoOpsetImportVersion any more. May be simpler to inline the internal function here.